### PR TITLE
common-tweaks layer

### PR DESCRIPTION
**Pull request by StreakyCobra**
Allow users to easily enable/disable non-trivial tweaks.

### CODE REVIEW

File: layers/+config-files/common-tweaks/README.org [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+To use this contribution add it to your =~/.spacemacs=:
+
+#+begin_src emacs-lisp
+  (setq-default dotspacemacs-configuration-layers '(common-tweaks))
+#+end_src
+
+This layer has no effect until some tweaks are enable through variables:
+
+#+begin_src emacs-lisp
+(setq-default dotspacemacs-configuration-layers '(common-tweaks
```
Comment (TheBB) [inline layers/+config-files/common-tweaks/README.org]: You're missing a set of parens here.

**Comment by TheBB** [comment]
Is there any reason why we need a complicated macro and not just a bunch of conditionals?

File: layers/+config-files/common-tweaks/README.org (hunk 1 above)
Comment (StreakyCobra) [inline layers/+config-files/common-tweaks/README.org]: Right.

**Comment by StreakyCobra** [comment]
> Is there any reason why we need a complicated macro and not just a bunch of conditionals?

Hum… It's not really a complicated one (except maybe the loader trick). Advantages are:

- It clearly separates different parts of the tweaks (Functions, tweaks, loading the code at the right moment, description)

- All functions and lines related to a tweak are grouped in the same entity, so it's easier to:

  - Keep the related code grouped together
  - Delete a tweak without forgetting some parts/functions of the tweak
  - Temporary disabling all related code with a flag `:disable t` in one go

- Avoid boilerplate code, like:

  - Defining a variable with the same name as the tweak
  - Print a message about enabled tweaks when `--debug-init` is passed
  - Checking the variable for activating the tweak or not
  - Synchronizing the documentation of the tweak with the documentation of the variable

- It will be easier to add a general functionality about tweak in only one place (whatever it is).

The macro is not really difficult to understand and quite straightforward (just grouping all parts with a `(progn …`, if we omit the `:loader` trick). You just need the `:tweak` property to be set, all other being optional and should be used only if needed.

I see many advantages for the code structure and its cleanness, without big drawbacks. Of course if it is really a problem, it's not difficult for me to change this.

**Comment by StreakyCobra** [comment]
Let me know here if there are other tweaks that can be useful to spacemacs users.

**Comment by justbur** [comment]
For `counsel-find-file`
```elisp
(setq counsel-find-file-ignore-regexp "\\(?:\\`[#.]\\)\\|\\(?:[#~]\\'\\)")
```
Ignores any file or directory starting with a dot unless the user starts the search string with a dot.

**Comment by StreakyCobra** [comment]
@justbur Thanks, done :+1:

**Comment by StreakyCobra** [comment]
To WIP or not to WIP, that is the question… I still have improvements to do, but I suppose it's ready to merge to have first users feedback.

**Comment by fbergroth** [comment]
Not sure if it belongs here, but I'd like to see
- restore windows after quitting ediff
- quit ediff without prompting

**Comment by StreakyCobra** [comment]
It may be. Do you already have snippets?

**Comment by fbergroth** [comment]
Not any good ones :-)

**Comment by fbergroth** [comment]
@StreakyCobra 
For restoring window layout:
``` el
(add-hook 'ediff-after-quit-hook-internal 'winner-undo)
```
Slightly modified your example for quitting ediff without prompt:
``` el
(defun ct//no-confirm (origfunc &rest args)
  (cl-letf (((symbol-function 'y-or-n-p) (lambda (&rest args) t)))
    (apply origfunc args)))

(advice-add 'ediff-quit :around 'ct//no-confirm)
```

**Comment by StreakyCobra** [comment]
Thx, I'll add this later this week :-)

**Comment by TheBB** [comment]
`(add-hook 'ediff-after-quit-hook-internal 'winner-undo)`

This at least should be a default, and in any case is not nontrivial.

**Comment by nixmaniack** [comment]
Yes, restoration of windows after ediff quit should be a default!

**Comment by StreakyCobra** [comment]
> This at least should be a default, and in any case is not nontrivial.

Agreed.

We also have to discuss what belongs to `common-tweaks` and what belongs to the layers.

**Comment by sooheon** [comment]
Everything from 115 to 122 here should be a tweak, imo: https://github.com/syl20bnr/spacemacs/blob/develop/layers/+distribution/spacemacs-bootstrap/packages.el#L115

There is some discussion about the >gv tweaks in #5273, and I think the J/K move line tweaks fall under the same umbrella.

**Comment by StreakyCobra** [comment]
@sooheon The link referencing the precise commit so we can find the specified lines if the file change later:

https://github.com/syl20bnr/spacemacs/blob/7f95d250dbea53028f24309cbc77469a3f0aa08c/layers/%2Bdistribution/spacemacs-bootstrap/packages.el#L115-L122

-----

Concerning the status of this PR, I'm not sure if it should be its own layer like it I'm proposing here, or if the tweaks should be proposed in the corresponding layer directly. I remember someone pointed this out and it was discussed, but I can't find out where :/

The idea was that if we propose the tweaks directly in the layers, then people just have to read the README of the corresponding layer and already know how to customize it. With a `common-tweaks` layer people have first to know about its existence, and then knows there are some tweaks regarding a given functionality before being able to use it.

**Comment by fbergroth** [comment]
> I remember someone pointed this out and it was discussed, but I can't find out where :/

@StreakyCobra The link referencing the precise message :smile: 
https://gitter.im/syl20bnr/spacemacs?at=56942381d739f50a3602f788

**Comment by StreakyCobra** [comment]
@fbergroth :+1: Nice thanks :-)

**Comment by sooheon** [comment]
Hm, especially given that now even spacemacs itself is broken up into layers, it makes even more sense.

**Comment by d12frosted** [comment]
@syl20bnr Is it really in `Merging...` state? 😸

**Comment by syl20bnr** [comment]
I'm still thinking about it but I lean toward dispatching the tweaks where they belongs. If a tweak is worthy it should be an option in its corresponding layer IMO.

**Comment by StreakyCobra** [comment]
:point_up: Agreed

**Pull request closed without merging**